presto-hive/src/main/java/io/prestosql/plugin/hive/SyncPartitionMetadataProcedure.java

```diff
@@ -0,0 +1,224 @@
+        }
+    }
+
+    private List<FileStatus> listDirectory(FileSystem fileSystem, FileStatus current, List<Column> partitionColumns, int depth)
+    {
+        if (depth == 0) {
+            return ImmutableList.of(current);
+        }
+
+        try {
+            return Stream.of(fileSystem.listStatus(current.getPath()))
+                    .filter(fileStatus -> validatePath(fileSystem, fileStatus, partitionColumns.get(partitionColumns.size() - depth)))
+                    .flatMap(directory -> listDirectory(fileSystem, directory, partitionColumns, depth - 1).stream())
+                    .collect(toImmutableList());
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private boolean validatePath(FileSystem fileSystem, FileStatus file, Column column)
+    {
+        try {
+            Path path = file.getPath();
+            String prefix = column.getName() + '=';
+            return fileSystem.isDirectory(path) && path.getName().startsWith(prefix);
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    // calculate relative complement of set b with respect to set a
+    private Set<String> difference(List<String> a, List<String> b)
+    {
+        return Sets.difference(newHashSet(a), newHashSet(b));
+    }
+
+    private void addPartitions(
+            SemiTransactionalHiveMetastore metastore,
+            ConnectorSession session,
+            Table table,
+            Path location,
+            Set<String> partitions)
+    {
+        partitions.forEach(name -> metastore.addPartition(
+                session,
+                table.getDatabaseName(),
+                table.getTableName(),
+                buildPartitionObject(session, table, name, new Path(location, name)),
+                new Path(location, name),
+                PartitionStatistics.empty()));
+    }
+
+    private void dropPartitions(
+            SemiTransactionalHiveMetastore metastore,
+            ConnectorSession session,
+            Table table,
+            Set<String> partitions)
+    {
+        partitions.forEach(name -> metastore.dropPartition(
+                session,
+                table.getDatabaseName(),
+                table.getTableName(),
+                extractPartitionValues(name)));
+    }
+
+    private Partition buildPartitionObject(ConnectorSession session, Table table, String partitionName, Path targetPath)
+    {
+        return Partition.builder()
+                .setDatabaseName(table.getDatabaseName())
+                .setTableName(table.getTableName())
+                .setColumns(table.getDataColumns())
+                .setValues(extractPartitionValues(partitionName))
+                .setParameters(ImmutableMap.of(PRESTO_QUERY_ID_NAME, session.getQueryId()))
+                .withStorage(storage -> storage
+                        .setStorageFormat(isRespectTableFormat(session) ?
```

Comment:
I don't think this is the right behavior. Respect table format controls what format Presto will use for new partitions that it writes. No matter which format is chosen, the partition data will match since Presto writes it in that format.

For adding partitions, the data already exists, but we don't know the format. The expected use case is that you would create an external table pointing at existing data, while specifying the format of the existing data, then call this procedure. Using the session format would be very unexpected.

We could also add an optional `format` option for the procedure to specify the format. (as a future enhancement if anyone needs it)